### DISCUSSION TITLE
--complicated-parser (both CLI and GUI) issue

### DISCUSSION THREAD

**Discussion by protofolius** [comictagger/comictagger#511] [Q&A]
I am currently in the process of adding metadata to my comics collection , which has been managed up to now by only folders and filestructures . Metadata is the way to go so I have some work on my hands . ;)

My filestructure is as follows : {series} #{issue} {title} ({year})
95% are .cbr files

Using the simple parser , the {series} , {issue} and {year} perfectly get inputted into the proper fields (both GUI and CLI) .
But I need to use the complicated parser in order to get the very important {title} part too .
Now this is where things get nasty .

The complicated parser seems to handle an {issue} part with only numbers just fine .
However I have used letters in my filenames for "special" editions .
A few {issue} examples -->

#P01 pocket editions
#F01 familiy book editions
#S01 special editions
#X01 X-rated editions

with obviously the last 2 digits incrementing the more issues the series has .

**The GUI with complicated parser** is unable to handle these files .
It imports everything in the Series field , detaches the # from the issue field and attaches it to the first series part . Like this -->   
{series}# {issue} {title} .  ({year}) gets imported correctly .
So the Series field gets completely unusable data inside , and {issue} {title} field stay empty . year field gets populated .

**The CLI command with complicated parser** is unable to handle these files too .
command used
`ComicTagger -g -s -f --complicated-parser --overwrite -t cr -i -v "filename.cbr"`

It crashes with the following error and imports no metadata at all .

```
comictagger | ERROR | CLI mode failed
Traceback (most recent call last):
  File "comictaggerlib\main.py", line 218, in main
  File "comictaggerlib\cli.py", line 182, in run
  File "comictaggerlib\cli.py", line 590, in process_file_cli
  File "comictaggerlib\cli.py", line 343, in save
AttributeError: 'Namespace' object has no attribute 'runtime_assume_issue_one'
```

Using auto tag function is no option since the results for me are awful (my books are not English language and maybe 25% gets tagged out of which 50% is incorrect and it often changes my titles to French language and so on) . Also , auto tag tends to crash the program a lot (using the latest beta 1.6.0 a4 , tried both on windows 10 and 11) .

So now my question :
in order to get the metatag fields properly populated with data from my filenames , what is the best approach here ?
What could be a proper workaround ? Maybe even with mass renaming all my files in a different way so Comictagger can handle the import of all fields properly . 
Suggestions (VERY) welcome :) 👍 💯 👍

**Comment by lordwelch** [comment]
So the complicated parser is called that because it does some weird things sometimes, but it should work better _most_ of the time. Specifically it trips up on this because of the letter after the `#` I haven't seen any letters start an issue number so it gets categorized differently.

Currently the complicated parser can _try_ to parse out these fields:
```python
alternate: str
annual: bool
archive: str
c2c: bool
fcbd: bool
issue: str
issue_count: str
publisher: str
remainder: str
series: str
title: str
volume: str
volume_count: str
year: str
```

The original parser can try to parse these fields:
```python
series: str
volume: str
year: str
issue_count: str
remainder: str
issue: str
```

I could possibly make it add those as an alternate number but that may cause issues on other peoples naming schemes so I would want it to be very specific to try and reduce that if possible.

Now both parsers ignore anything after two dashes `--` they also treat anything in parenthesis special so you can add parenthesis around your special edition numbers and it will move that to the remainder variable which shows up in tags under scan_info and in the GUI on the credits tab at the bottom it is labeled Scan Info

The [wiki] has some explanation of the filename parsing rules for the original parser. The complicated parser can parse all of the examples listed and most of the time better than the original.

To fix the stability issues I'm going to put out an alpha release today, it won't have any other sources bundled in (like I promised in #501) but it is much more stable than the current alpha

[wiki]: https://github.com/comictagger/comictagger/wiki/UserGuide#filename-parsing

**Comment by protofolius** [comment]
So with my filestructure being {series} #{issue} {title} ({year})
Where the {issue} part consists not always of digits only ...

What would you suggest I do in order to get both Title and Issue field populated on import/first write of CR tag ?

Start using a different naming scheme alltogether for the "special edition issues" ? If so , a suggestion what would works best with a combination of:
Comictagger / Auto-tag Comicvine / Codex Comic book server ?

OR

Wait for a new version of comictagger that is able to handle an import with a combination of Letters and Digits in the #{issue} field ?

**Reply by lordwelch** [reply]
What does comicvine, GCD or Metron list for the issue number? ComicTagger can't auto-tag without an issue number that matches what is on ComicVine, or in the future on another service, so no matter what ComicTagger does it can't tag it automatically if it doesn't match.

I don't think codex cares at all what you have in the issue number field. ComicTagger doesn't technically care either it's just the filename parser that doesn't.

Specifically it is because you start with a letter after the `#` if you move it to the other side it will parse it properly. Again it's only this way because I've never seen any issue numbers that actually start with a letter. I've tested changing the parser to accept issue numbers that start with a letter and it hasn't broken any of the other file name parsing so I may include it in a few weeks.

If you just want to use Codex then you can parse out the issue separately (eg using powershell, I can give you a script later if you would like) and pass it in with -m "issue=$issue" and codex will read it just fine, but if you want auto-tagging to work you are at the mercy of ComicVine and whatever data they have in their database.

**Comment by protofolius** [comment]
The ones I have found on comicvine , they have either listed in another language (French) , with titles in different language , or with completely bogus and incorrect tagged editions .
So if I use auto tag , and UNtick the overwrite function , it might sometimes find the French edition BUT:
once it proposes me the match and I accept it , it overwrites both the title and publisher which I had set before . 
With both being wrong (French publisher and French title) .

So since Comicvine overall has very little of the comic editions in my collection properly listed (like I said my comics are not in English) , I rather keep relying on my own naming schedule if possible .

SoTherefore it would be very cool if some future version of ComicTagger would not trip over {issue} tags that carry both letters and numbers .

At least to me they make a lot of sense . Some examples are
#B01 biography/best of
#C01 compact editions
#E01 entrtainment/puzzle editions
#F01 familiy book editions
#J01 jubileum (anniversary) editions
#P01 pocket editions
#N01 newly brought  out/restyled editions
#O01 old editions (or oblong format)
#S01 special editions
#X01 X-rated editions

You get the idea .


-Would also be cool not having to use the CLI in order to mass apply CR tags based on filename . Since mass removal IS possible in the GUI .
-Would also be cool to be able to apply a CR tag to a new .cbr file and at the same time apply some additional metadata with it . Now it needs 2 runs in order to do this (1-write cr tag to cbr file 2-apply extra metadata info) . Not sure if this is possible at all though .
-Itunes in my opinion has a great mass tag GUI implementation for music files . But ok their software is not being written bij 1 person obviously . :)

**Reply by lordwelch** [reply]
GCD provides dumps of their database and I'm going to use it to investigate some issue numbers, check back in a few weeks and I'll see what I decide on.

All of the overwrite/clear options are kind of a mess right now so it's understandable if it doesn't do what you would expect and it needs to be fixed.
Adding a checkbox in the auto-tag window for whether it does an online search or not is a good idea, could you open an issue for that?

On the CLI tags get applied in least to greatest priority `-f` filename data -> `-t` tag data -> `-m` metadata supplied on the commandline. You can use all of these options at the same time. For example if it parses the `series`, `title` and `issue` from the filename and then you tell it to read the CR tags and it has the `issue` and `title` set and then you manually set the `title` on the commandline you end up with:
series: (from filename)
issue: (from CR Tags)
title: (from commandline)

Mass editing #23 has been requested long before I even knew they published digital comics (I only picked up this project in 2021)
If someone is able to prototype what that might look like using qt designer and provide a ui file (like is used here https://github.com/comictagger/comictagger/tree/40bd3d5bb8befa06e1078627f3ecb38a0a2246b0/comictaggerlib/ui) I don't have the time or drive to make a GUI bulk edit right now.

**Reply by lordwelch** [reply]
So apparently I don't know the difference between a few hours and a few weeks. Try out this PR https://github.com/comictagger/comictagger/pull/518 here are instructions on how to download the pre-built packages https://github.com/comictagger/comictagger/wiki/downloading-packages-from-pull-requests

**Comment by protofolius** [comment]
`On the CLI tags get applied in least to greatest priority -f filename data -> -t tag data -> -m metadata supplied on the commandline. You can use all of these options at the same time. For example if it parses the series, title and issue from the filename and then you tell it to read the CR tags and it has the issue and title set and then you manually set the title on the commandline you end up with:
series: (from filename)
issue: (from CR Tags)
title: (from commandline)`

Confirmed . This actually works . Will save me quite some time !


I opened an issue for the checkmark auto-tag idea (online/offline) in order to mass tag based on current form data (imported from filename) . I hope I did it the right way . Thank you !

**Comment by protofolius** [comment]
If you want to test the complicated parser on title imports , I can suggest the Michel Vaillant series of comics:
https://www.stripweb.be/R-titels-michel-vaillant.asp

With some of the titles:
F-3000
De $spon$or
100.000.000$ Voor Steve Warson
24 Uur Onder Invloed
F.1 In Oproer
Km 357
Het Spook Van De 24 Uur
Nr 13 Is Aan De Start
De 8e Man
Mach 1 Voor Steve Warson
Paris-Dakar
300 Km/h Door Parijs
Rendez-Vous in Macao
De Ridders van Köningsfeld


All fun :)


I noticed something else with the complicated tagger:
Every time there is a title with a digit , it strips the digit away .  
For example "The 5 Circles" , it always makes "The Circles" out of it .

A title of "De 4de Macht" , it is stripped and becomes "De" .

It also always removes the space before a & or a + sign . So a title of Drystan & Esyllt in the filename becomes Drystan& Esyllt for some reason .

I also mass imported some files and it turns out series "Favorieten Reeks" in the filename sometimes gets out of the parser as "Favorietenreeks" in 1 words , sometimes it comes at as "Favorieten Reeks" . So yes that is somewhat strange .

Maybe some things can be improved upon , I don't know .
